source/common/upstream/original_dst_cluster.cc

```diff
@@ -40,6 +41,26 @@ OriginalDstCluster::LoadBalancer::LoadBalancer(PrioritySet& priority_set, Cluste
 
 HostConstSharedPtr OriginalDstCluster::LoadBalancer::chooseHost(LoadBalancerContext* context) {
   if (context) {
+    // Check if request has a override host header set, if directly return the host using it.
+    const Http::HeaderMap* downstream_headers = context->downStreamHeaders();
+    if (downstream_headers &&
+        downstream_headers->get(Http::Headers::get().OriginalDstHostOverride) != nullptr) {
+      const std::string& request_override_host =
+          downstream_headers->get(Http::Headers::get().OriginalDstHostOverride)->value().c_str();
+      try {
+        Network::Address::InstanceConstSharedPtr overridehost_ip_port(
+            Network::Utility::parseInternetAddressAndPort(request_override_host, false));
+        ENVOY_LOG(debug, "Using request override host {}.", request_override_host);
+        return HostSharedPtr{new HostImpl(
+            info_, info_->name() + overridehost_ip_port->asString(),
+            std::move(overridehost_ip_port), envoy::api::v2::core::Metadata::default_instance(), 1,
+            envoy::api::v2::core::Locality().default_instance(),
+            envoy::api::v2::endpoint::Endpoint::HealthCheckConfig().default_instance())};
```

Comment:
Dont you need the logic below where the host is added to host_map_ (i.e. check if the host already exists in the existing cache)? We use that to garbage collect unused hosts over time.

Reply:
i am thinking we should use the host for one single request and leave it. In our use case, further requests wont come to the same host again. But do we loose any thing if we do not add it the map?

Reply:
I see no reason to not use all the existing logic per @rshriram. I haven't used the original dst cluster, but, I think my posting to the main thread and allowing it to be garbage collected, we can see the hosts in the clusters output? This is very useful for debugging. Can we factor the function so that we first decide what host to add (or not), and then go on to do the addition?